### PR TITLE
[OptimizeInstructions] Simplify floating points with NaN on right side

### PR DESCRIPTION
**Pull request by MaxGraey**
```rust
x + nan     ->    nan'
x - nan     ->    nan'
x * nan     ->    nan'
x / nan     ->    nan'
min(x, nan) ->    nan'
max(x, nan) ->    nan'
```
where `nan'` is canonicalized `nan` of rhs
```rust
x != nan    ->    1
x == nan    ->    0
x >= nan    ->    0
x <= nan    ->    0
x >  nan    ->    0
x <  nan    ->    0
```

### CODE REVIEW

**Comment by MaxGraey** [comment]
`Fuzzed: 18427`

**Comment by MaxGraey** [comment]
@kripken Could you also look at this PR?

**Comment by MaxGraey** [comment]
@tlively could you take a look please?

File: src/literal.h [hunk 1]
```diff
@@ -706,6 +707,40 @@ struct GCData {
   GCData(HeapType type, Literals values) : type(type), values(values) {}
 };
 
+// Wasm has nondeterministic rules for NaN propagation in some operations. For
```
Comment (kripken) [inline src/literal.h:710]: Maybe move the comment and declaration here, but the body can remain in the C++ I think - I don't believe this must be inlined for speed.

File: src/passes/OptimizeInstructions.cpp [hunk 2]
```diff
@@ -3597,6 +3597,57 @@ struct OptimizeInstructions
         }
       }
     }
+    {
+      // copysign(x, +C)   ==>   abs(x)
+      // copysign(x, -C)   ==>   neg(abs(x))
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:3602]: This one seems slower - two operations instead of one.

File: src/passes/OptimizeInstructions.cpp [hunk 3]
```diff
@@ -3597,6 +3597,57 @@ struct OptimizeInstructions
         }
       }
     }
+    {
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:3600]: The title of this PR makes sense for the NaN opts, but the copysign stuff seems unrelated. It could be in a separate PR?

File: src/literal.h (hunk 1 above)
Comment (kripken) [reply src/literal.h:710]: (though it might be worth measuring)

File: src/passes/OptimizeInstructions.cpp (hunk 2 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:3602]: No, In many architectures nabs is single instruction. See [this](https://github.com/bytecodealliance/wasmtime/issues/4803). While `copysign` required 3-4 instructions.

File: src/passes/OptimizeInstructions.cpp (hunk 3 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:3600]: Yes, you can do a separate PR. I decided in one, so they are connected. In fact, copysign is a more general, but separate case. I can put copysign rules in a separate PR

Comment (kripken) [reply src/passes/OptimizeInstructions.cpp:3600]: Maybe I just don't see the connection. But the code is separate, tests separate, etc., so it seems like separate PRs are best.

Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:3600]: Sure! Will do

File: src/passes/OptimizeInstructions.cpp (hunk 2 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:3602]: I think sooner or later they will optimize this as well as single instruction. I will create an issue for v8 for this. But even if web engines don't optimize yet it should still be faster than `copysign` or at least with the same speed. While it will definitely have a smaller size since we get rid of the constant (-8 bytes)

Comment (kripken) [reply src/passes/OptimizeInstructions.cpp:3602]: why would it be the same speed, if they don't optimize it? surely copysign takes as many cycles as abs or neg?

Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:3602]: That's how copysign lowered on x64 for example:
https://github.com/bytecodealliance/wasmtime/blob/2986f6b0fffd576a2ba88a1fc6cb2db959a1f9e8/cranelift/filetests/filetests/isa/x64/fcopysign.clif#L33

So it's 1 `movabsq` +  2 `movdqa` + 1 `andpd` + 1 `orpd`

Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:3602]: For example with AVX it could looks like this:
```asm
mov 0x7fffffff, %r10d
vmovd %r10d, %xmm3
vpternlogd $0xe4, %xmm3, %xmm1, %xmm0 
```

Comment (kripken) [reply src/passes/OptimizeInstructions.cpp:3602]: Interesting, but I don't know enough about this. Does x64 really not have a CPU instruction for copysign? A wasm principle has always been to add wasm instructions only for things that have CPU instructions that correspond to them, in major architectures at least.

File: src/literal.h (hunk 1 above)
Comment (MaxGraey) [reply src/literal.h:709]: done!

File: src/passes/OptimizeInstructions.cpp (hunk 3 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:3600]: done!

File: src/passes/OptimizeInstructions.cpp [hunk 4]
```diff
@@ -3597,6 +3597,40 @@ struct OptimizeInstructions
...
+      //   x <=> NaN   ==>   0
+      //   x op  NaN'  ==>   NaN',  iff `op` != `copysign` and `x` != C
+      Const* c;
+      Binary* bin;
+      Expression* x;
+      if (matches(curr, binary(&bin, pure(&x), fval(&c))) &&
+          std::isnan(c->value.getFloat()) && !x->is<Const>() &&
+          bin->op != getBinary(x->type, CopySign)) {
+        if (bin->isRelational()) {
+          // reuse "c" (nan) constant
```
[the remark this answers is not in the document]
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:3611]: That's done in line below:
https://github.com/WebAssembly/binaryen/pull/4985/files#diff-d03831e8b5cc4d080b5884abbf6d6d2776449ca53b3655a2cd5a07e57d6a11c5R3612

File: src/passes/OptimizeInstructions.cpp [hunk 5]
```diff
@@ -3597,6 +3597,40 @@ struct OptimizeInstructions
...
     }
+    {
+      //   x !=  NaN   ==>   1
+      //   x <=> NaN   ==>   0
+      //   x op  NaN'  ==>   NaN',  iff `op` != `copysign` and `x` != C
+      Const* c;
+      Binary* bin;
+      Expression* x;
+      if (matches(curr, binary(&bin, pure(&x), fval(&c))) &&
+          std::isnan(c->value.getFloat()) && !x->is<Const>() &&
```
[the remark this answers is not in the document]
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:3608]: I usually skipped the constant to the right side for the floating point because there could be NaN, which was better to handle in propagate constant pass. I think since OptimizeInstruction now uses the same utility for nan-canonicalization we can remove this restriction.

File: src/passes/OptimizeInstructions.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:3611]: No problem

**Review by kripken — APPROVED** [review]
Thanks, lgtm. Does this need any more fuzzing before landing - have there been any non-style changes since you fuzzed?

**Comment by MaxGraey** [comment]
I'll do a little fuzz testing now to make sure that removing the constant check does not break the vm comparisons

**Comment by MaxGraey** [comment]
Ok, refuzzed until `3871` iterations
I guess it's ready to merge

**Comment by kripken** [comment]
Ok, sounds good.
